File: src/main/scala/leetcode/Solution242.scala
package leetcode

class Solution242 {


  /**
   * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
   *
   * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
   * typically using all the original letters exactly once.
   *
   * @param s
   * @param t
   * @return
   */
  def isAnagram(s: String, t: String): Boolean = {
    val sArr = s.toCharArray.toSeq.sorted
    if (s.length != t.length) return false
    val mapOfSStr = sArr.foldLeft(scala.collection.mutable.Map.empty[Char, Int])((map, item) => map += (item -> s.count(_ == item))).toMap
    val unmatchedChar = mapOfSStr.dropWhile(tuple2 => {
      val (currentChar, count) = tuple2
      count == t.count(_ == currentChar)
    })
    unmatchedChar.isEmpty
  }

}

File: java-solution/src/main/java/leetcode/Solution242.java
package leetcode;


public class Solution242 {
    /**
     * Given two strings s and t, return true if t is an anagram of s, and false otherwise.
     *
     * An Anagram is a word or phrase formed by rearranging the letters of a different word or phrase,
     * typically using all the original letters exactly once.
     * @param s
     * @param t
     * @return
     */
    public boolean isAnagram(String s, String t) {
        if (s.length() != t.length()) return false;
        char[] tComposedChars = new char[t.length()];
        char[] tRemainingChar = t.toCharArray();
        //fill the data
        for (int i = 0; i < s.length(); i++) {
            char currentChar = s.charAt(i);
            if(t.indexOf(currentChar) < 0) return false;
            tComposedChars[i] = tRemainingChar[t.indexOf(currentChar)];
//            tRemainingChar[t.indexOf(currentChar)] = ' ';
        }
        System.out.printf("%s %s", s, String.valueOf(tComposedChars));
        return s.equals(String.valueOf(tComposedChars));
    }
}
